sympy/tensor/tensor.py

```diff
@@ -1769,7 +1772,22 @@ def __call__(self, *indices, **kw_args):
         A(a, -b)
 
         """
-        tensor = Tensor(self, indices, **kw_args)
+
+        updated_indices = []
+        for idx, typ in zip(indices, self.index_types):
+            if isinstance(idx, str):
+                idx = idx.strip().replace(" ", "")
+                if idx.startswith('-'):
+                    updated_indices.append(TensorIndex(idx[1:], typ,
+                                           is_up=False))
+                else:
+                    updated_indices.append(TensorIndex(idx, typ))
+            else:
+                updated_indices.append(idx)
+
+        updated_indices += indices[len(updated_indices):]
```

Comment:
What's the point of this line? Isn't the previous for-loop not supposed to loop over all indices?

Reply:
No, it does not. It will loop over all indices only if `len(indices) == len(self.index_types)`. If there are more indices than types, then there are some indices which are not looped over (because the string cannot be converted using the type), and this line puts them back in the list. This should not happen, but checking that there is the correct number of indices and triggering the error is handled by `Tensor`.

Reply:
I'm not convinced this is the best way to handle this. Maybe adding a check that the lengths of indices and index types are the same would be better?

Reply:
Also... the code is a bit too hard to read compared to what it's supposed to do.

Reply:
> I'm not convinced this is the best way to handle this. Maybe adding a check that the lengths of indices and index types are the same would be better?

I don't see the point of making twice the checks for consistency of the indices. The point of my code is to convert string indices to proper indices when possible (that is, for the first `n` indices given, where `n` is the number of types). Before my modification there was no check of the length at this point of the code: why should it be added now if it was not deemed necessary before and if it's checked in any case in `Tensor`?

> Also... the code is a bit too hard to read compared to what it's supposed to do.

I don't see why: I am just reading how many indices have been converted, and taking the remaining ones to put them after. I am happy to have a suggestion for something simpler then.

Reply:
is the case of mismatched numbers of indices being tested?

Reply:
I don't understand the question. Tested in my code or in `Tensor`? In the first case, no as explained above, in the second case, yes (and this raises an error).

Reply:
Yes, just want to make sure that there is a test checking that the a wrong indices number raises an exception.